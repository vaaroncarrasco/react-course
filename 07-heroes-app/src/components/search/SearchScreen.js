import React, { useMemo } from 'react';
import queryString from 'query-string';
import { HeroCard } from '../heroes/HeroCard';
import { useForm } from '../../hooks/useForm';
import { useLocation } from 'react-router-dom';
import { getHeroesByName } from '../selectors/getHeroesByName';

// Add SearchScreen
// implement submit of form w/ handleSearch and cl text
// use cusotm hooks from remote repo

// * React Forms
// 1) Use useForm() custom hook that uses useState()
// 2) set initialState -> useForm({ inputName: inputValue }) -> useForm({ searchText: '' })
// 3) Destruct state to get input value -> const { searchText } = formValues;
// 4) handleSubmit = (e) => { e.preventDefault(); }
// 5) HTML <form onSubmit={ handleSubmit }
// 6) <input name="searchText" value={ searchText } onChange={ handleInputChange }
// 7) <btn type="submit" -> to call handleSubmit

export const SearchScreen = ({ history }) => {

    // * getting ?q=query
    const location = useLocation(); // location.search -> ?q=batman

    // if q undefined -> = ''
    const { q = '' } = queryString.parse( location.search ); // batman

    const [ formValues, handleInputChange ] = useForm({
        // same as <input name="searchText"
        searchText: q
    });

    const { searchText } = formValues;

    const heroesFiltered = useMemo(() => getHeroesByName( q ), [ q ]);

    const handleSearch = (e) => {
        e.preventDefault();
        history.push(`?q=${ searchText }`);
    }

    return (
        <div>
            <h1>Search Screen</h1>
            <hr />

            <div className="row">

                <div className="col-5">
                    <h4>Search Form</h4>
                    <hr />

                    <form onSubmit={ handleSearch }>
                        <input
                            name="searchText"
                            value={ searchText }
                            onChange={ handleInputChange }
                            type="text"
                            placeholder="Find your hero"
                            className="form-control"
                            autoComplete="off"
                        />

                        <button
                            type="submit"
                            className="btn m-1 btn-block btn-outline-primary"
                        >
                            Search...
                        </button>
                    </form>

                </div>

                <div className="col-7">

                    <h4> Results </h4>
                    <hr />

                    {
                        (q === '')
                            &&
                            <div className="alert alert-info">
                                Search a hero
                            </div>
                    }

                    {
                        (q !== '' && heroesFiltered.length === 0)
                            &&
                            <div className="alert alert-danger">
                                There's no hero named { q }
                            </div>
                    }

                    {
                        heroesFiltered.map( hero => (
                            <HeroCard
                                key={ hero.id }
                                { ...hero }
                            />
                        ))
                    }

                </div>

            </div>

        </div>
    )
}
