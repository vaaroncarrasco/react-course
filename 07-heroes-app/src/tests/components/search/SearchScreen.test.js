import { mount } from 'enzyme'
import React from 'react'
import { MemoryRouter, Route } from 'react-router'
import { SearchScreen } from '../../../components/search/SearchScreen';

describe('testing <SearchScreen />', () => {

    test('should display with default values', () => {

        const wrapper = mount(
            // ? Testing search thru URL params, import MemoryRouter, Route
            <MemoryRouter initialEntries={['/search']}>
                <Route path="/search" component={ SearchScreen } />
            </MemoryRouter>
        );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.alert-info').text().trim() ).toBe('Search a hero');

    });

    test('should show batman and input with query string', () => {
        const wrapper = mount(
            // ? Testing search thru URL params, import MemoryRouter, Route
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path="/search" component={ SearchScreen } />
            </MemoryRouter>
        );
        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('input').prop('value') ).toBe('batman');
    });

    test('should error box if hero not found', () => {
        const wrapper = mount(
            // ? Testing search thru URL params, import MemoryRouter, Route
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <Route path="/search" component={ SearchScreen } />
            </MemoryRouter>
        );
        expect( wrapper.find('.alert-danger').exists() ).toBe(true);
    });

    test('should call history.push()', () => {

        const history = {
            push: jest.fn()
        }

        const wrapper = mount(
            // ? Testing search thru URL params, import MemoryRouter, Route
            <MemoryRouter initialEntries={['/search?q=batman123']}>

                <Route
                    path="/search"
                    // * Passing route's functional components props
                    component={ () => <SearchScreen history={ history } /> }
                />

            </MemoryRouter>
        );


        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText',
                value: 'batman'
            }
        });

        // ? Simulating form submit with preventDefault prop sent thru
        wrapper.find('form').prop('onSubmit')( { preventDefault(){} } );

        expect( history.push ).toHaveBeenCalledWith('?q=batman');

    });

});