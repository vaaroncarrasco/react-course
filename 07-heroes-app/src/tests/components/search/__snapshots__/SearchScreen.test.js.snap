// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing <SearchScreen /> should display with default values 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
         Results 
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;

exports[`testing <SearchScreen /> should show batman and input with query string 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
         Results 
      </h4>
      <hr />
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                 
                Batman
                 
              </h5>
              <p
                className="card-text"
              >
                 
                Bruce Wayne
                 
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  Detective Comics #27
                   
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                MÃ¡s...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
